### ISSUE TITLE
Sentinel-terminated assign is lossy

### ISSUE THREAD

**Issue by pascaldekloe** [ziglang/zig#17668]
### Zig Version

0.12.0-dev.1128+1bbe52107

### Steps to Reproduce and Observed Behavior

Assignment of a sentinel-terminated value to a non-sentinel terminated value loses data without warning.

```zig
const std = @import("std");

test "tricky" {
    // allocate sentinel-terminated
    var s = try std.testing.allocator.dupeZ(u8, "foo");

    // lose the sentinel with assignment
    const T = struct { s: []const u8 };
    var t = T{ .s = s };

    // [BUG] free won't work due size mismatch
    std.testing.allocator.free(t.s);
}
```

### Expected Behavior

Zig tends to prevent mistakes with explicitness rather than barricading unsafe options. The strategy works quite well, even for people who ain't used to such freedom. I think Zig should require `@as` (or something like that) when assignment loses the sentinel.

```zig
    var t = T{ .s = @as([]const u8, s) };
```

**Comment by rohlem** [comment]
`@as` usually only specifies a type that the value automatically coerces to, so requiring it somewhere would be without precedent afaik.
However, we could require slicing syntax `s[0..]` for the conversion (which doesn't include the sentinel unless it is specified, as in `s[0..:0]`).
That is also more terse than the required type repetition in `@as`.

**Comment by pascaldekloe** [comment]
The `s[0..:]` syntax is crystal clear. 💎😍 I'm happy to provide an implementation this week if there's "enough" support for the concept.

**Comment by InKryption** [comment]
I'll just note, this same issue applies to alignment:
```zig
const slice: []u8 = try allocator.alignedAlloc(u8, 4, 32);
defer allocator.free(slice);
```

**Comment by pascaldekloe** [comment]
Ouch! That Zig lizzard is trying to kill us all. 😂 Slicing would be good on allignment loss too. I'll see what I can do to make this boring.

**Comment by pascaldekloe** [comment]
The language specification does [explicitly permit](https://ziglang.org/documentation/0.11.0/#Type-Coercion-Stricter-Qualification) “align - bigger to smaller alignment is allowed” now in subsection “Type Coercion: Stricter Qualification”. Can we agree that this was a mistake, and that lowering the alignment is a lossy operation?

**Comment by InKryption** [comment]
> Slicing would be good on allignment loss too

Still not quite a full-proof solution:
```zig
const ptr: *align(1) u32 = try allocator.create(u32);
defer allocator.destroy();
```
And were there ever to be a method added like `createAligned`, this would be worsened.

It does make me think it might be worth re-thinking how we allow down-casting pointer attributes; perhaps it being the same as the integer promotion model (`u[X]` -> `u[X + 1]`) isn't a completely solid idea, being that it loses valuable information silently.

**Comment by pascaldekloe** [comment]
Perhaps as follows @InKryption?

```diff
--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -28837,10 +28837,14 @@ const InMemoryCoercionResult = union(enum) {
                 break;
             },
             .ptr_sentinel => |sentinel| {
-                if (sentinel.actual.toIntern() != .unreachable_value) {
+                if (sentinel.actual.toIntern() != .unreachable_value and sentinel.wanted.toIntern() != .unreachable_value) {
                     try sema.errNote(block, src, msg, "pointer sentinel '{}' cannot cast into pointer sentinel '{}'", .{
                         sentinel.actual.fmtValue(sentinel.ty, mod), sentinel.wanted.fmtValue(sentinel.ty, mod),
                     });
+                } else if (sentinel.actual.toIntern() != .unreachable_value) {
+                    try sema.errNote(block, src, msg, "pointer sentinel '{}' lost; cut-off with …[0..:] or …[*:] instead", .{
+                        sentinel.actual.fmtValue(sentinel.ty, mod),
+                    });
                 } else {
                     try sema.errNote(block, src, msg, "destination pointer requires '{}' sentinel", .{
                         sentinel.wanted.fmtValue(sentinel.ty, mod),
@@ -29444,8 +29448,9 @@ fn coerceInMemoryAllowedPtrs(
         } };
     }
```

**Comment by dweiller** [comment]
I don't really see this as a problem with coercion, but rather as a finicky aspect of the allocator API that one needs to be aware of when reinterpreting (i.e. casting/coercing) allocated memory. I don't think changing how coercion works is really going to help here, as a call to `free`/`destroy` is often not going to be right next to an allocation or coercion, so a more explicit change in the type won't be clearly related to a usage with the allocator API.

Moreover the same issues could be seen for any coercion as all coercions lose information that could be used by an API. I think this is just a potential issue that any API that uses precise type information to change runtime behaviour is going to have to think about, and users of these APIs need to be aware of their requirements.

**Comment by andrewrk** [comment]
Loss of type information during a coercion is common and expected. For example:
* losing pointer mutability
* losing pointer alignment
* an error set to anyerror
* a pointer to anyopaque

**Comment by pascaldekloe** [comment]
None of these issues are obvious. When I read an assignment then I don't expect interesting things to happen there. I certainly don't want to check the field type of every assignment just to get the correct context. Write for the reader!

```
    t.s = s[0..:];
    t.p = p[*:];
```

It is really hard to follow how anyone could be against such ☝️ clarity @andrewrk. What are we about to lose with explicitness?

Yes Allocator.free accepts errors in silence which is a serious problem on in't own, yet it is just one demonstration of why the current behaviour is dangerous. There are many more scenario in which silent loss will bite 💯 guaranteed.

**Comment by Pyrolistical** [comment]
I agree lost of type information during explicit coercion is acceptable, but this is during an innocent looking assignment.

Here is another repo case:
```zig
const std = @import("std");

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    const oops_forgot_sentinel: []u8 = try allocator.allocSentinel(u8, 10, 0); // imo should compile error here
    defer {
        // correctly throws error(gpa): Allocation size 11 bytes does not match free size 10.
        allocator.free(oops_forgot_sentinel);
    }
}
```

This error caught at compile time instead of runtime.
